### ISSUE TITLE
Whether or not the `drain_collect` lint suggestion is an improvement is context dependent

### ISSUE THREAD

**Issue by udoprog** [rust-lang/rust-clippy#12410]
### Description

The [drain_collect](https://rust-lang.github.io/rust-clippy/master/index.html#/drain_collect) lint will currently flag code like this:

```
fn method(values: &mut Vec<u8>) -> Vec<u8> {
    values.drain(..).collect()
}
```

With:

```
warning: you seem to be trying to move all elements into a new `Vec`
 --> src/main.rs:2:5
  |
2 |     values.drain(..).collect()
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider using `mem::take`: `std::mem::take(values)`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#drain_collect
  = note: `#[warn(clippy::drain_collect)]` on by default
```

Since this is marked as a perf hint, it's expected that the change has a different behavior than the original code. However, whether or not the proposed change is an improvement is in my mind too context dependent to be a default suggestion.

If for example `values` is a collection that is re-used explicitly to make use of its max allocated size, it might in the long run lead to more allocations swapping it than draining it would have.

### Version

```text
rustc 1.76.0 (07dca489a 2024-02-04)
binary: rustc
commit-hash: 07dca489ac2d933c78d3c5158e3f43beefeb02ce
commit-date: 2024-02-04
host: x86_64-pc-windows-msvc
release: 1.76.0
LLVM version: 17.0.6
```

**Comment by blyxyas** [comment]
Hi, thanks for the report, could you make a reproducible example where the lint suggestion is not optimal, so that we know what cases to target.

**Comment by udoprog** [comment]
Sure:

```rust
use core::mem::take;

fn main() {
    let mut vec = Vec::new();
    let mut vecs = Vec::<Vec<usize>>::new();
    
    let mut allocations = 0;

    for _ in 0..10 {
        for n in 0..100 {
            let a = vec.capacity();

            vec.push(n);

            if a != vec.capacity() {
                allocations += 1;
            }
        }

        // Note: collecting into a new vector is one allocation.
        vecs.push(vec.drain(..).collect());
        allocations += 1;

        // vecs.push(take(&mut vec));
    }

    dbg!(allocations, vecs.len());
}
```

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=331e21d98e5303396357016f8a891e40).

The version which uses `take` as clippy recommends performs 60 allocations, while the version which drain collects performs 16.

**Comment by blyxyas** [comment]
Honestly, I'm not sure how to approach the problem. We'll talk about this in the next meeting, you can check our Zulip stream and provide more information. The next meeting will be on

**Comment by asayers** [comment]
I sometimes use this pattern: I push onto a long-lived large-capacity `Vec`, and then sometimes `drain(..).collect()` it to emit the current contents.  This is nice because (A) `push()` never triggers reallocation, and (B) the `Vec`s you produce have the exact right size.  When clippy shows me this lint I "fix" it by collecting into a `Box<[T]>` instead of a `Vec`.  So you _could_ say clippy has made me improve my code, slightly, maybe, but it wasn't by following the lint's advice.
